File: 백준알고리즘/BFS(너비 우선 탐색)/DFS와 BFS.py
import sys
input = sys.stdin.readline

N, M, V = map(int, input().split())
graph = [[0] * (N + 1) for _ in range(N + 1)]

for _ in range(M):
    m1, m2 = map(int, input().split())
    graph[m1][m2] = graph[m2][m1] = 1


def bfs():
    visited = list()
    need_visit = list()
    need_visit.append(V)

    while need_visit:
        node = need_visit.pop(0)

        if node not in visited:
            visited.append(node)
            for i in range(N + 1):
                if graph[node][i] == 1:
                    need_visit.append(i)

    return visited


dfs_visited = [0] * (N + 1)


def dfs(n):
    dfs_visited[n] = 1
    print(n, end=' ')
    for i in range(1, N + 1):
        if dfs_visited[i] == 0 and graph[n][i] == 1:
            dfs(i)


dfs(V)
print()
print(*bfs())

File: 백준알고리즘/BFS(너비 우선 탐색)/javascript/DFS와 BFS.js
const fs = require('fs');
const stdin = fs.readFileSync('/dev/stdin').toString().split('\n');

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

const [N, M, V] = input()
  .split(' ')
  .map(n => +n);

const graph = new Array(N + 1).fill(0).map(() => new Array(N + 1).fill(0));

for (let i = 0; i < M; i++) {
  const [m1, m2] = input()
    .split(' ')
    .map(n => +n);

  graph[m1][m2] = graph[m2][m1] = 1;
}

const dfs_answer = [];
dfs_visited = new Array(N + 1).fill(0);

function dfs(n) {
  dfs_visited[n] = 1;
  dfs_answer.push(n);

  for (let i = 0; i < N + 1; i++) {
    if (dfs_visited[i] === 0 && graph[n][i] === 1) dfs(i);
  }
}
dfs(V);
console.log(...dfs_answer);

function bfs(v) {
  const visited = [];
  const need_visited = [v];

  while (need_visited.length !== 0) {
    const node = need_visited.shift();

    if (!visited.includes(node)) {
      visited.push(node);

      for (let i = 0; i < N + 1; i++) {
        if (graph[node][i] === 1) need_visited.push(i);
      }
    }
  }
  return visited;
}

console.log(...bfs(V));
